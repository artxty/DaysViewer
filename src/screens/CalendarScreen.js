import React, { useState, useRef, useContext } from 'react';
import { Portal } from 'react-native-portalize';
import { View } from 'react-native';
import { NavigationEvents } from 'react-navigation';

import RatingModal from '../components/RatingModal';
import Calendar from '../components/Calendar';
import useDaysStorage from '../storage/daysStorageHook';
import { SettingsContext } from '../storage/settingsContext';
import appInfoStorage from '../storage/appInfoStorage';

const EMPTY_DAY = (id) => ({ id, rating: 0, note: '' });

export default function CalendarScreen() {
  const { usePastMonthsRange } = useContext(SettingsContext);
  const [pastMonthsRange] = usePastMonthsRange;

  const modalRef = useRef(null);
  const [modalData, setModalData] = useState(EMPTY_DAY('2020-01-01'));

  const [visibleMonths, setVisibleMonths] = useState([]);
  const [days, setDaysToStorage, refetchHookData] = useDaysStorage(visibleMonths);

  function handleOnDayPress(id) {
    setModalData(days[id] !== null ? { ...days[id], id } : EMPTY_DAY(id));
    modalRef.current.open();
  }

  async function handleDaySaving({ id, rating, note }) {
    if (!(rating === 0 && note.length === 0)) {
      await setDaysToStorage({ [id]: { rating, note } });
    }
    modalRef.current.close();
  }

  return (
    <View>
      <NavigationEvents
        onWillFocus={async () => {
          const doesNeedUpdate = await appInfoStorage.get('doesCalendarNeedUpdate');
          if (doesNeedUpdate) {
            refetchHookData();
            await appInfoStorage.set('doesCalendarNeedUpdate', false);
          }
        }}
      />
      <Portal>
        <RatingModal
          ref={modalRef}
          dayData={modalData}
          onSave={handleDaySaving}
        />
      </Portal>
      <Calendar
        pastMonthsRange={pastMonthsRange}
        days={days}
        onScroll={(months) => setVisibleMonths(months)}
        onDayPress={handleOnDayPress}
      />
    </View>
  );
}

CalendarScreen.navigationOptions = {
  header: null,
};
