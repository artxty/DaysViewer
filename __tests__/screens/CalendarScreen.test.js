import React from 'react';
import renderer from 'react-test-renderer';

import CalendarScreen from '../../src/screens/CalendarScreen';
import useDaysStorage from '../../src/storage/daysStorageHook';

jest.mock('react-native-portalize', () => ({
  Portal: 'Portal',
}));

jest.mock('react-navigation', () => ({
  NavigationEvents: 'NavigationEvents',
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: () => ({
    usePastMonthsRange: [3],
  }),
}));

jest.mock('../../src/components/RatingModal', () => 'RatingModal');
jest.mock('../../src/components/Calendar', () => 'Calendar');

const dayId = '2020-01-01';
const days = {
  [dayId]: {
    rating: 1,
    note: 'Note123',
  },
};

jest.mock('../../src/storage/daysStorageHook', () => jest.fn());
useDaysStorage.mockImplementation(() => [
  days,
  jest.fn(),
]);

describe('CalendarScreen', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CalendarScreen />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  // TODO: should call modal.close when onSave
  // TODO: should call RatingModal.open when onDayPress
  // TODO: passes dayData param to RatingModal when onDayPress
});
