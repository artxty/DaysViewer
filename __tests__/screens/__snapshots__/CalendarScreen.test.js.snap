// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarScreen renders correctly 1`] = `
<View>
  <NavigationEvents
    onWillFocus={[Function]}
  />
  <Portal>
    <RatingModal
      dayData={
        Object {
          "id": "2020-01-01",
          "note": "",
          "rating": 0,
        }
      }
      onSave={[Function]}
    />
  </Portal>
  <Calendar
    days={
      Object {
        "2020-01-01": Object {
          "note": "Note123",
          "rating": 1,
        },
      }
    }
    onDayPress={[Function]}
    onScroll={[Function]}
    pastMonthsRange={3}
  />
</View>
`;
